Add unified binary proposal

Hoverbear opened the pull request:
cc @zhouqiang-cl 

This is a proposal to unify the TiKV binaries into a single `tikv` binary. This would be part of a major version change since it would have some backwards incompatibilities.

zhouqiang-cl commented on text/2019-11-06-unified-binary.md:
```diff
@@ -0,0 +1,149 @@
+
+### Smaller build graph
+
+Running `cargo +nightly build -Z timings --release` allows for the measurement
```
how many time it takes to build

Hoverbear replied:
That's on the next line :)

zhouqiang-cl commented on text/2019-11-06-unified-binary.md:
````diff
@@ -0,0 +1,149 @@
+    unsafe-recover        Unsafely recover the cluster when the majority replicas are failed
+```
+
+Moving `tikv-server` to a `tikv server --flags` and keeping all the `tikv-ctl *`
````
we also need to consider compatibility when upgrade

Hoverbear replied:
Yes, this is why we create aliases :)

zhangjinpeng1987 replied:
Luckily tikv-ctl and tikv-server don't have command line args conflict(should check again more carefully).
I agree to unified those binaries.

zhouqiang-cl commented:
@zhangjinpeng1987 @AndreMouche @siddontang @winkyao @nolouch How did you think about this proposal

Hoverbear commented:
Friendly ping

Hoverbear commented:
Ping @zhangjinpeng1987 @AndreMouche @siddontang @winkyao @nolouch

zhangjinpeng1987 commented:
I prefer sub command like `tikv-server ctl xxxx` and `tikv-server server xxx`.

Hoverbear commented:
@zhangjinpeng1987 So you'd prever to see `tikv ctl xxx` instead of `tikv xxx`, so that the `tikv` command can only be called with a subcommand?

zhangjinpeng1987 commented:
@Hoverbear Yes.

Hoverbear commented:
@zhangjinpeng1987 That's definitely a great option, let's see if others have some opinion on this?

The original reason I suggested this is for two reasons:

1. It's not idiomatic to start daemons and services like TiKV directly from command line. They should be run via something like `systemctl enable tikv --now` or `rc update add tikv` or otherwise using an init system. So most users would very rarely, call` tikv server`, the majority of the time they would likely call `tikv ctl`.
2. This would create a big complexity difference between `tikv server`, which just accepts some flags, and `tikv ctl` which has many subcommands and options. The downside of this is that `tikv server` kind of gets lost in the subcommands, but as noted, most users should be using provides init service scripts.

Here's an interaction with `ctl` as a separate command:

```bash
$ tikv --help
TiKV (tikv) 
Release Version:   4.0.1-alpha
Git Commit Hash:   e54b2eb06d1560b64cf23160f28fa74ff7a394ee
Git Commit Branch: rpm
UTC Build Time:    2019-12-13 08:44:16
Rust Version:      rustc 1.39.0-nightly (c6e9c76c5 2019-09-04)

A distributed transactional key-value store.

USAGE:
    tikv [FLAGS] [OPTIONS] [SUBCOMMAND]

FLAGS:
    -h, --help                    Prints help information
    -V, --version                 Prints version information

SUBCOMMANDS:
    ctl                           Interact with a TiKV instance.
    server                        Begin a TiKV Server Daemon.

---

$ tikv ctl --help
TiKV Control (tikv) 
Release Version:   4.0.1-alpha
Git Commit Hash:   e54b2eb06d1560b64cf23160f28fa74ff7a394ee
Git Commit Branch: rpm
UTC Build Time:    2019-12-13 08:44:16
Rust Version:      rustc 1.39.0-nightly (c6e9c76c5 2019-09-04)

A distributed transactional key-value store.


USAGE:
    tikv ctl [FLAGS] [OPTIONS] [SUBCOMMAND]

FLAGS:
    -h, --help                    Prints help information
        --skip-paranoid-checks    Skip paranoid checks when open rocksdb
    -V, --version                 Prints version information

OPTIONS:
        --ca-path <ca_path>              Set the CA certificate path
        --cert-path <cert_path>          Set the certificate path
        --cipher-file <cipher_file>      set cipher file path
        --config <config>                Set the config for rocksdb
        --db <db>                        Set the rocksdb path
        --decode <decode>                Decode a key in escaped format
        --encode <encode>                Encode a key in escaped format
        --to-hex <escaped-to-hex>        Convert an escaped key to hex key
        --to-escaped <hex-to-escaped>    Convert a hex key to escaped key
        --host <host>                    Set the remote host
        --key-path <key_path>            Set the private key path
        --pd <pd>                        Set the address of pd
        --raftdb <raftdb>                Set the raft rocksdb path

SUBCOMMANDS:
    bad-regions           Get all regions with corrupt raft
    cluster               Print the cluster id
    compact               Compact a column family in a specified range
    compact-cluster       Compact the whole cluster in a specified range in one or more column families
    consistency-check     Force a consistency-check for a specified region
    diff                  Calculate difference of region keys from different dbs
    dump-snap-meta        Dump snapshot meta file
    fail                  Inject failures to TiKV and recovery
    help                  Prints this message or the help of the given subcommand(s)
    metrics               Print the metrics
    modify-tikv-config    Modify tikv config, eg. tikv-ctl --host ip:port modify-tikv-config -m kvdb -n
                          default.disable_auto_compactions -v true
    mvcc                  Print the mvcc value
    print                 Print the raw value
    raft                  Print a raft log entry
    random-hex            Generate random bytes with specified length and print as hex
    raw-scan              Print all raw keys in the range
    recover-mvcc          Recover mvcc data on one node by deleting corrupted keys
    recreate-region       Recreate a region with given metadata, but alloc new id for it
    region-properties     Show region properties
    scan                  Print the range db range
    size                  Print region size
    split-region          Split the region
    store                 Print the store id
    tombstone             Set some regions on the node to tombstone by manual
    unsafe-recover        Unsafely recover the cluster when the majority replicas are failed

---

$ tikv server --help
TiKV (tikv)
Release Version:   4.0.1-alpha
Git Commit Hash:   e54b2eb06d1560b64cf23160f28fa74ff7a394ee
Git Commit Branch: rpm
UTC Build Time:    2019-12-13 08:44:16
Rust Version:      rustc 1.39.0-nightly (c6e9c76c5 2019-09-04)

A distributed transactional key-value store.

USAGE:
    tikv server [FLAGS] [OPTIONS]

FLAGS:
        --config-check           
            Check config file validity and exit

    -h, --help                   
            Prints help information

        --print-sample-config    
            Print a sample config to stdout

    -V, --version                
            Prints version information


OPTIONS:
    -A, --addr <IP:PORT>              
            Set the listening address

        --advertise-addr <IP:PORT>    
            Set the advertise listening address for client communication

        --capacity <CAPACITY>         
            Set the store capacity to use. If not set, use entire partition

    -C, --config <FILE>               
            Set the configuration file

    -s, --data-dir <PATH>             
            Set the directory used to store data

        --labels <KEY=VALUE>...       
            Set the server labels. Uses `,` to separate kv pairs, like `zone=cn,disk=ssd`

    -f, --log-file <FILE>             
            Set the log file path. If not set, logs will output to stderr

    -L, --log-level <LEVEL>           
            Set the log level [possible values: trace, debug, info, warn, warning, error, critical]

        --metrics-addr <IP:PORT>      
            Sets push address to the Prometheus Pushgateway, leaves it empty will disable Prometheus push

        --pd-endpoints <PD_URL>...    
            Set the PD endpoints to use. Use `,` to separate multiple PDs

        --status-addr <IP:PORT>       
            Set the HTTP listening address for the status report service
```


And now as the default:

```bash
$ tikv --help
TiKV (tikv) 
Release Version:   4.0.1-alpha
Git Commit Hash:   e54b2eb06d1560b64cf23160f28fa74ff7a394ee
Git Commit Branch: rpm
UTC Build Time:    2019-12-13 08:44:16
Rust Version:      rustc 1.39.0-nightly (c6e9c76c5 2019-09-04)

A distributed transactional key-value store.


USAGE:
    tikv [FLAGS] [OPTIONS] [SUBCOMMAND]

FLAGS:
    -h, --help                    Prints help information
        --skip-paranoid-checks    Skip paranoid checks when open rocksdb
    -V, --version                 Prints version information

OPTIONS:
        --ca-path <ca_path>              Set the CA certificate path
        --cert-path <cert_path>          Set the certificate path
        --cipher-file <cipher_file>      set cipher file path
        --config <config>                Set the config for rocksdb
        --db <db>                        Set the rocksdb path
        --decode <decode>                Decode a key in escaped format
        --encode <encode>                Encode a key in escaped format
        --to-hex <escaped-to-hex>        Convert an escaped key to hex key
        --to-escaped <hex-to-escaped>    Convert a hex key to escaped key
        --host <host>                    Set the remote host
        --key-path <key_path>            Set the private key path
        --pd <pd>                        Set the address of pd
        --raftdb <raftdb>                Set the raft rocksdb path

SUBCOMMANDS:
    bad-regions           Get all regions with corrupt raft
    cluster               Print the cluster id
    compact               Compact a column family in a specified range
    compact-cluster       Compact the whole cluster in a specified range in one or more column families
    consistency-check     Force a consistency-check for a specified region
    diff                  Calculate difference of region keys from different dbs
    dump-snap-meta        Dump snapshot meta file
    fail                  Inject failures to TiKV and recovery
    help                  Prints this message or the help of the given subcommand(s)
    metrics               Print the metrics
    modify-tikv-config    Modify tikv config, eg. tikv-ctl --host ip:port modify-tikv-config -m kvdb -n
                          default.disable_auto_compactions -v true
    mvcc                  Print the mvcc value
    print                 Print the raw value
    raft                  Print a raft log entry
    random-hex            Generate random bytes with specified length and print as hex
    raw-scan              Print all raw keys in the range
    recover-mvcc          Recover mvcc data on one node by deleting corrupted keys
    recreate-region       Recreate a region with given metadata, but alloc new id for it
    region-properties     Show region properties
    scan                  Print the range db range
    server                Start a TiKV Daemon

    size                  Print region size
    split-region          Split the region
    store                 Print the store id
    tombstone             Set some regions on the node to tombstone by manual
    unsafe-recover        Unsafely recover the cluster when the majority replicas are failed

---

$ tikv server --help
TiKV (tikv)
Release Version:   4.0.1-alpha
Git Commit Hash:   e54b2eb06d1560b64cf23160f28fa74ff7a394ee
Git Commit Branch: rpm
UTC Build Time:    2019-12-13 08:44:16
Rust Version:      rustc 1.39.0-nightly (c6e9c76c5 2019-09-04)

A distributed transactional key-value store.

USAGE:
    tikv server [FLAGS] [OPTIONS]

FLAGS:
        --config-check           
            Check config file validity and exit

    -h, --help                   
            Prints help information

        --print-sample-config    
            Print a sample config to stdout

    -V, --version                
            Prints version information


OPTIONS:
    -A, --addr <IP:PORT>              
            Set the listening address

        --advertise-addr <IP:PORT>    
            Set the advertise listening address for client communication

        --capacity <CAPACITY>         
            Set the store capacity to use. If not set, use entire partition

    -C, --config <FILE>               
            Set the configuration file

    -s, --data-dir <PATH>             
            Set the directory used to store data

        --labels <KEY=VALUE>...       
            Set the server labels. Uses `,` to separate kv pairs, like `zone=cn,disk=ssd`

    -f, --log-file <FILE>             
            Set the log file path. If not set, logs will output to stderr

    -L, --log-level <LEVEL>           
            Set the log level [possible values: trace, debug, info, warn, warning, error, critical]

        --metrics-addr <IP:PORT>      
            Sets push address to the Prometheus Pushgateway, leaves it empty will disable Prometheus push

        --pd-endpoints <PD_URL>...    
            Set the PD endpoints to use. Use `,` to separate multiple PDs

        --status-addr <IP:PORT>       
            Set the HTTP listening address for the status report service
```

Hoverbear commented:
Closing. I am not longer involved with the project and have no stake in this merging. Feel free to create a new PR.
